### ISSUE TITLE
Label cache not saved properly if train/val labels are not in split folders

### ISSUE THREAD

**Issue by JfranekAC** [ultralytics/ultralytics#6659]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

I've been trying out training a detection model in YOLOv8 and I've noticed that the label cache got recalculated every time even if the dataset didn't change. I've been using dataset structure:
```
dataset/
├── images/
├── labels/
├── train.txt
└── val.txt
```

I've found out that in `YOLODataset.get_labels()` the path to cache file is defined as `cache_path = Path(self.label_files[0]).parent.with_suffix('.cache')` In this case the path for both the training and validation datasets resolves to `labels.cache` so the validation cache rewrites the training cache. Which means the cache hash check always fails for both datasets and label cache has to be recalculated.

### Minimal Reproducible Example

```
from ultralytics import YOLO

model=YOLO('yolov8n.pt')

model.train(data="*path_to_dataset*/cache_example/data.yaml", epochs=3)
```

Example dataset:
[cache_example.zip](https://github.com/ultralytics/ultralytics/files/13501234/cache_example.zip)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!
